### ISSUE TITLE
Windows SDL 1.x: SDL surface as child window of parent window, rather than directly as window, for async move/resize/menu interaction

### ISSUE THREAD

**Issue by joncampbell123** [joncampbell123/dosbox-x#472]
Problem:
- ~~Windows UI move/resize (SYSCOMMAND SC_MOVE/SC_SIZE) and menu dropdown interaction is blocking~~ SOLVED
- ~~SDL's message pump is blocked during window resize, move, or handling a menu dropdown~~ SOLVED
- ~~Emulator stops during resize, move, and menu interaction~~ SOLVED
- ~~This is due to undocumented behavior regarding DefWindowProc() within a window callback, which upon certain non-client area interaction, will block the caller and run a message pump within itself to respond to the UI.~~ SOLVED
- ~~Windows are owned by a thread, and there are problems with one thread controlling or managing other thread's windows.~~ SOLVED

Solution:
- ~~SDL_SetVideoMode() first call should start the "parent window thread" which creates the parent window (with titlebar, min/max/restore, sysmenu, and windows menus). The parent window thread will run it's own message pump.~~
- ~~When the parent thread has initialized the parent window, and signalled readiness, the main thread in SDL_SetVideoMode() should then create a borderless child window (WS_CHILD) which then acts as it does now, becoming the main display area for the SDL video surface.~~
- ~~The parent window will handle window resize events and resize the child window to match it's client area. The child window will react to resize events as it does now, and send SDL resize events in the main thread.~~
- ~~If the main thread calls to resize the window (SDL_SetVideoMode) the child window will SendMessage() WM_USER events to the parent thread window to indicate that it wants the parent to size up so the client area is the desired size. SendMessage() is involved to make it synchronous between child and parent, which gives the parent the chance to act on the request and return to the child window the final size of the client area.~~
- ~~Another message: whether the child window wants to enable sizing (SDL_RESIZABLE), which will indicate to the parent whether to enable the maximize buttons.~~
- ~~Another message: whether the parent window is maximized (and therefore SDL_SetVideoMode has no effect on the window size)~~
- ~~Another message: WM_COMMAND messages from the parent to the child (the menu bar), and synchronous messages to clear/set the menu bar.~~
- ~~Another message to set the window title of the parent window from the child window (and main thread) so SDL library calls to do that can work.~~

Results:
- ~~Interacting with the menu bar will not cause the emulator to pause~~ CHECK
- ~~Dragging the window will not cause the emulator to pause~~ CHECK
- ~~Resizing will not cause the emulator to pause, and the emulator can react to the resizing as it happens.~~ CHECK

Other potential issues:
- ~~Need to make sure this change does not break fullscreen modes~~ CHECK
- ~~Need to make sure this doesn't break joystick input, mouse input, mouse capture, touchscreen input, etc.~~ CHECK
- ~~Need to make sure SDL_SetVideoMode() starts parent window thread only ONCE.~~ CHECK
- ~~Need to make sure SDL_Quit() signals the parent window thread to stop, and the parent window thread safely destroys the window as part of shutdown, and that SQL_Quit() does not return until the parent window thread has finished.~~ CHECK
- ~~Need to make sure that both parent window thread, parent window, child window, SDL video surface, etc. are all shut down and de-initialized properly at SDL_Quit().~~ CHECK
- ~~Make sure that SDL_Init() does not re-init the parent window thread and cause it to start again when already started.~~ CHECK
- ~~DOSBox code in some cases forces a SDL state reset by calling SDL_Quit() then SDL_Init() and SDL_SetVideoMode(), make sure that does not break this scheme.~~ CHECK
- ~~Upon entering fullscreen mode, the child window needs to remove itself from the parent and become the fullscreen code (so as not to break fullscreen code already in SDL 1.x). The parent window should be hidden after the child window is de-parented.~~ CHECK
- ~~Upon leaving fullscreen mode, the child window needs to put itself back into the parent window and show the parent window again. The parent window should be shown (without activating or bringing it to the front), then the child window re-parented into it.~~ CHECK
- ~~SDL syswm structure needs to be extended to contain the HWND of both the parent and child window.~~ CHECK
- ~~DOSBox-X GetHWND() needs to return the parent window (menu bar, MessageBox(), GetOpenFileName, etc. model operations), but some calls to GetHWND() should obtain the child window instead (Direct3D output) where relevant.~~ CHECK
- ~~What if the child window requests a resize or other operation while the parent window is busy with a drag, resize, or menu item operation? Ideally (and probably true), is that the use of SendMessage() will block until the drag/resize/menu operation is finished. No resizing the window out from under the user.~~ CHECK

Other thoughts:
- ~~Could other Windows-specific hacks be rolled into the SDL 1.x library as well?~~ YES. DOSBox-X's hack adding the menu bar, for example, has been incorporated into the SDL 1.x library.

**Comment by joncampbell123** [comment]
The only remaining issue is that fullscreen mode is not yet reliable.

It MOSTLY works, but sometimes SDL enters fullscreen mode with only a partial region displayed, that often corresponds to the original window size.

I also noticed that sometimes, when resizing the window, resize events don't make it. Not often, but sometimes.

**Comment by joncampbell123** [comment]
Because of the massive changes made to accomplish this goal, I will wait to merge it to develop, and then master, and leave it in the develop-win-sdl1-async-hack-201802 branch for further testing.

I am concerned that while this happens to work so far on Windows 10, it may break in subtle ways on Windows 7, or worse yet, on Windows XP where I do not have the ability to compile test the code.

**Comment by joncampbell123** [comment]
Another bug noted is that if you ALT+TAB away, and then ALT+TAB back, focus is not on the child window, and keyboard input to the DOS game does not work because of that.

**Comment by joncampbell123** [comment]
3Dfx emulation (from my testing) seems to work well with the async mode now. Code has been added to prevent anything that would disrupt it including bringing up the mapper and "GUI" or hiding/showing the menu bar.

**Comment by joncampbell123** [comment]
Target date to merge into "develop" branch (and then into master) is March 1st. I will push the date back further if other issues are not resolved by then.

**Comment by joncampbell123** [comment]
Reminder: Async behavior will be merged into develop and then master, March 1st. Async will become default behavior.

**Comment by joncampbell123** [comment]
Reminder: Async behavior will be merged into develop and then master, March 1st. Async will become default behavior on Windows.

**Comment by joncampbell123** [comment]
Reminder: Async behavior will be merged into develop and then master, tonight. Async will become default and only behavior on Windows.

**Comment by joncampbell123** [comment]
This has been done.
